src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java

```diff
@@ -1112,15 +1116,21 @@ private void rewriteDependencyReducedPomIfWeHaveReduction(
                     w.close();
                 }
 
-                ProjectBuildingRequest projectBuildingRequest =
-                        new DefaultProjectBuildingRequest(session.getProjectBuildingRequest());
-                projectBuildingRequest.setLocalRepository(session.getLocalRepository());
-                projectBuildingRequest.setRemoteRepositories(project.getRemoteArtifactRepositories());
+                // Lock critical section to fix MSHADE-467
+                try {
```

Comment:
A global lock looks wrong, shouldnt we lock on the project (we can store a concurrenthashmap of project/lock in the session to avoid synchronized for the future if we enable virtual threads a day)

Reply:
I tried, it is not enough. Feel free to suggest something better, but after you verified that it works, please.

Reply:
BTW, feel free to commit on top of the PR, we can always squash some commits later, as long as we keep the IT separate from the fix.

![image](https://github.com/apache/maven-shade-plugin/assets/1537384/140dd01b-d84d-4100-a42f-7085a7f05c00)

Reply:
```
diff --git a/src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java b/src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
index 3a7eaaa..a8475d9 100644
--- a/src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
+++ b/src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
@@ -35,8 +35,6 @@ import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
 
 import org.apache.maven.RepositoryUtils;
 import org.apache.maven.artifact.Artifact;
@@ -1050,8 +1048,6 @@ public class ShadeMojo extends AbstractMojo {
         rewriteDependencyReducedPomIfWeHaveReduction(dependencies, modified, transitiveDeps, model);
     }
 
-    private static final Lock LOCK = new ReentrantLock();
-
     private void rewriteDependencyReducedPomIfWeHaveReduction(
             List<Dependency> dependencies, boolean modified, List<Dependency> transitiveDeps, Model model)
             throws IOException, ProjectBuildingException, DependencyGraphBuilderException {
@@ -1116,9 +1112,7 @@ public class ShadeMojo extends AbstractMojo {
                     w.close();
                 }
 
-                // Lock critical section to fix MSHADE-467
-                try {
-                    LOCK.lock();
+                synchronized (session.getProjectBuildingRequest()) { // Lock critical section to fix MSHADE-467
                     ProjectBuildingRequest projectBuildingRequest =
                             new DefaultProjectBuildingRequest(session.getProjectBuildingRequest());
                     projectBuildingRequest.setLocalRepository(session.getLocalRepository());
@@ -1128,8 +1122,6 @@ public class ShadeMojo extends AbstractMojo {
 
                     getLog().debug("updateExcludesInDeps()");
                     modified = updateExcludesInDeps(result.getProject(), dependencies, transitiveDeps);
-                } finally {
-                    LOCK.unlock();
                 }
             }
 

```

can you give a try to that and check it is sufficient for you please? should fix your issue without putting a global bottleneck in the mojo (previous comment is still accurate, no strong opinion on lock vs synchronized as of today, just want a bottleneck per project to avoid a bottleneck in complex projects).

Reply:
LGTM, thanks for the great input. I made sure your authorship was respected in the latest commit. The CI build on my fork is passing. Before merging, would you like me to squash our two solutions into one commit? I wwould prefer to keep both commits for reference and to see the code evolution. Should the necessity for a reentrant lock arise in the future, at least we have an example. We also have one already for what is generally preferable, like in this case. As I cannot merge myself, I leave it up to you to decide.

Reply:
Looks ok to me like that - if another commiter is ok then let's go ;)